ts/src/grpc.ts

```diff
@@ -130,53 +171,62 @@ export namespace grpc {
       transport = DefaultTransportFactory.getTransport();
     }
     transport({
+      debug: props.debug || false,
       url: `${props.host}/${methodDescriptor.service.serviceName}/${methodDescriptor.methodName}`,
       headers: requestHeaders,
       body: framedRequest,
       credentials: "",
       onHeaders: (headers: BrowserHeaders, status: number) => {
+        props.debug && debug("onHeaders", headers, status);
         responseHeaders = headers;
-        props.debug && console.debug("onHeaders", headers, status);
-        props.debug && console.debug("responseHeaders", JSON.stringify(responseHeaders, null, 2));
+        props.debug && debug("onHeaders.responseHeaders", JSON.stringify(responseHeaders, null, 2));
+        const code = httpStatusToCode(status);
+        props.debug && debug("onHeaders.code", code);
+        const gRPCMessage = headers.get("grpc-message");
+        props.debug && debug("onHeaders.gRPCMessage", gRPCMessage);
+        if (code !== Code.OK) {
+          rawOnError(code, gRPCMessage.length > 0 ? gRPCMessage[0] : "");
+          return;
+        }
+
         rawOnHeaders(headers);
       },
       onChunk: (chunkBytes: Uint8Array) => {
-        props.debug && console.debug("onChunk: ", chunkBytes);
-
         let data: Chunk[] = [];
         try {
           data = parser.parse(chunkBytes);
         } catch (e) {
-          props.onError(e);
+          props.debug && debug("onChunk.parsing error", e, e.message);
+          rawOnError(Code.Internal, e.message);
           return;
         }
 
         data.forEach((d: Chunk) => {
-          props.debug && console.debug("onChunk ", d);
-
           if (d.chunkType === ChunkType.MESSAGE) {
-            rawOnMessage(methodDescriptor.responseType.deserializeBinary(d.data!));
+            const deserialized = methodDescriptor.responseType.deserializeBinary(d.data!);
+            rawOnMessage(deserialized);
           } else if (d.chunkType === ChunkType.TRAILERS) {
+            props.debug && debug("onChunk.trailers", responseTrailers);
             responseTrailers = new BrowserHeaders(d.trailers);
           }
         });
       },
       onComplete: () => {
-        props.debug && console.debug("grpc.onComplete");
+        props.debug && debug("grpc.onComplete");
```

Comment:
what happens on resolution failures etc? is that gonna go through onComplete? If so, can you make sure we return a proper error here? and maybe add a comment as to why :)